File: infrastructure/cloudflare/src/onboarding/enumFieldValues.ts
/**
 * Allowed-value lists for the activity-3 fields typed ENUM. There is no
 * `allowed_values` column on `field_definition` - this stays a small, local
 * constant rather than a schema change, kept narrow to just the new
 * enum-typed fields introduced alongside it. Mirrors
 * `backend/.../domain/EnumFieldValues.java`.
 */
const SUFFIXES = new Set(['JR', 'SR', 'I', 'II', 'III', 'IV'])

const US_STATES = new Set([
  'AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'FL', 'GA',
  'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 'LA', 'ME', 'MD',
  'MA', 'MI', 'MN', 'MS', 'MO', 'MT', 'NE', 'NV', 'NH', 'NJ',
  'NM', 'NY', 'NC', 'ND', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC',
  'SD', 'TN', 'TX', 'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY',
  'DC', 'PR',
])

const MARITAL_STATUSES = new Set(['SINGLE', 'MARRIED', 'DIVORCED', 'WIDOWED'])

const CITIZENSHIP_STATUSES = new Set(['US_CITIZEN', 'RESIDENT_ALIEN', 'NON_RESIDENT_ALIEN'])

const EMPLOYMENT_STATUSES = new Set(['EMPLOYED', 'SELF_EMPLOYED', 'RETIRED', 'STUDENT', 'HOMEMAKER', 'UNEMPLOYED'])

/** Shared coarse-range buckets for income/net-worth/liquid-net-worth. */
const MONEY_RANGES = new Set([
  'UNDER_25K', 'FROM_25K_TO_50K', 'FROM_50K_TO_100K', 'FROM_100K_TO_200K', 'FROM_200K_TO_500K', 'OVER_500K',
])

const TAX_BRACKET_RANGES = new Set(['LOW', 'MODERATE', 'HIGH', 'HIGHEST'])

/** Simplified to single-select - no array/multi-select support in this schema's flat string values. */
const SOURCE_OF_FUNDS = new Set([
  'EMPLOYMENT_INCOME', 'INVESTMENTS', 'INHERITANCE', 'RETIREMENT_SAVINGS', 'BUSINESS_INCOME', 'OTHER',
])

const INVESTMENT_OBJECTIVES = new Set(['INCOME', 'GROWTH', 'GROWTH_AND_INCOME', 'SPECULATION', 'CAPITAL_PRESERVATION'])

const RISK_TOLERANCES = new Set(['CONSERVATIVE', 'MODERATE', 'AGGRESSIVE'])

/** Simplified to one overall experience level rather than a per-product breakdown. */
const INVESTMENT_EXPERIENCES = new Set(['NONE', 'LIMITED', 'GOOD', 'EXTENSIVE'])

const TIME_HORIZONS = new Set(['SHORT_TERM', 'MEDIUM_TERM', 'LONG_TERM'])

const DELIVERY_PREFERENCES = new Set(['E_DELIVERY', 'PAPER'])

const COST_BASIS_METHODS = new Set(['FIFO', 'LIFO', 'SPECIFIC_IDENTIFICATION', 'AVERAGE_COST'])

const BY_FIELD_KEY: Record<string, Set<string>> = {
  suffix: SUFFIXES,
  residentialState: US_STATES,
  mailingState: US_STATES,
  maritalStatus: MARITAL_STATUSES,
  citizenship: CITIZENSHIP_STATUSES,
  employmentStatus: EMPLOYMENT_STATUSES,
  annualIncomeRange: MONEY_RANGES,
  netWorthRange: MONEY_RANGES,
  liquidNetWorthRange: MONEY_RANGES,
  taxBracketRange: TAX_BRACKET_RANGES,
  sourceOfFunds: SOURCE_OF_FUNDS,
  investmentObjective: INVESTMENT_OBJECTIVES,
  riskTolerance: RISK_TOLERANCES,
  investmentExperience: INVESTMENT_EXPERIENCES,
  timeHorizon: TIME_HORIZONS,
  deliveryPreference: DELIVERY_PREFERENCES,
  costBasisMethod: COST_BASIS_METHODS,
}

/** Undefined for a field key with no enum allowlist. */
export function allowedValuesFor(fieldKey: string): Set<string> | undefined {
  return BY_FIELD_KEY[fieldKey]
}

File: backend/java-service/src/main/java/com/falcon/onboarding/domain/EnumFieldValues.java
package com.falcon.onboarding.domain;

import java.util.Map;
import java.util.Set;

/**
 * Allowed-value lists for the activity-3 fields typed {@link FieldDataType#ENUM}.
 * There is no {@code allowed_values} column on {@code field_definition} - this
 * stays a small, local constant rather than a schema change, kept narrow to
 * just the new enum-typed fields introduced alongside it.
 */
public final class EnumFieldValues {

    private static final Set<String> SUFFIXES = Set.of("JR", "SR", "I", "II", "III", "IV");

    private static final Set<String> US_STATES = Set.of(
            "AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DE", "FL", "GA",
            "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD",
            "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ",
            "NM", "NY", "NC", "ND", "OH", "OK", "OR", "PA", "RI", "SC",
            "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY",
            "DC", "PR");

    private static final Set<String> MARITAL_STATUSES = Set.of("SINGLE", "MARRIED", "DIVORCED", "WIDOWED");

    private static final Set<String> CITIZENSHIP_STATUSES = Set.of("US_CITIZEN", "RESIDENT_ALIEN", "NON_RESIDENT_ALIEN");

    private static final Set<String> EMPLOYMENT_STATUSES = Set.of(
            "EMPLOYED", "SELF_EMPLOYED", "RETIRED", "STUDENT", "HOMEMAKER", "UNEMPLOYED");

    /** Shared coarse-range buckets for income/net-worth/liquid-net-worth - "coarse synthetic ranges instead of exact sensitive financial values" per docs/06. */
    private static final Set<String> MONEY_RANGES = Set.of(
            "UNDER_25K", "FROM_25K_TO_50K", "FROM_50K_TO_100K", "FROM_100K_TO_200K", "FROM_200K_TO_500K", "OVER_500K");

    private static final Set<String> TAX_BRACKET_RANGES = Set.of("LOW", "MODERATE", "HIGH", "HIGHEST");

    /**
     * The reference brokerage's real form allows multiple sources of funds; simplified to a
     * single-select here since this codebase's field values are flat strings
     * with no array/multi-select support yet - a documented scope reduction.
     */
    private static final Set<String> SOURCE_OF_FUNDS = Set.of(
            "EMPLOYMENT_INCOME", "INVESTMENTS", "INHERITANCE", "RETIREMENT_SAVINGS", "BUSINESS_INCOME", "OTHER");

    private static final Set<String> INVESTMENT_OBJECTIVES = Set.of(
            "INCOME", "GROWTH", "GROWTH_AND_INCOME", "SPECULATION", "CAPITAL_PRESERVATION");

    private static final Set<String> RISK_TOLERANCES = Set.of("CONSERVATIVE", "MODERATE", "AGGRESSIVE");

    /** Simplified to one overall experience level rather than the reference brokerage's per-product (stocks/bonds/options/etc.) breakdown. */
    private static final Set<String> INVESTMENT_EXPERIENCES = Set.of("NONE", "LIMITED", "GOOD", "EXTENSIVE");

    private static final Set<String> TIME_HORIZONS = Set.of("SHORT_TERM", "MEDIUM_TERM", "LONG_TERM");

    private static final Set<String> DELIVERY_PREFERENCES = Set.of("E_DELIVERY", "PAPER");

    private static final Set<String> COST_BASIS_METHODS = Set.of(
            "FIFO", "LIFO", "SPECIFIC_IDENTIFICATION", "AVERAGE_COST");

    private static final Map<String, Set<String>> BY_FIELD_KEY = Map.ofEntries(
            Map.entry("suffix", SUFFIXES),
            Map.entry("residentialState", US_STATES),
            Map.entry("mailingState", US_STATES),
            Map.entry("maritalStatus", MARITAL_STATUSES),
            Map.entry("citizenship", CITIZENSHIP_STATUSES),
            Map.entry("employmentStatus", EMPLOYMENT_STATUSES),
            Map.entry("annualIncomeRange", MONEY_RANGES),
            Map.entry("netWorthRange", MONEY_RANGES),
            Map.entry("liquidNetWorthRange", MONEY_RANGES),
            Map.entry("taxBracketRange", TAX_BRACKET_RANGES),
            Map.entry("sourceOfFunds", SOURCE_OF_FUNDS),
            Map.entry("investmentObjective", INVESTMENT_OBJECTIVES),
            Map.entry("riskTolerance", RISK_TOLERANCES),
            Map.entry("investmentExperience", INVESTMENT_EXPERIENCES),
            Map.entry("timeHorizon", TIME_HORIZONS),
            Map.entry("deliveryPreference", DELIVERY_PREFERENCES),
            Map.entry("costBasisMethod", COST_BASIS_METHODS));

    private EnumFieldValues() {
    }

    /** Null for a field key with no enum allowlist (i.e. not an ENUM-typed field this class knows about). */
    public static Set<String> allowedValuesFor(String fieldKey) {
        return BY_FIELD_KEY.get(fieldKey);
    }
}
